pkg/authenticator/target_cluster_client.go

```diff
@@ -170,24 +176,46 @@ func (tcc *targetClusterClient) CreateClusterNamespace(ctx context.Context, clus
 		return fmt.Errorf("creating client for %s: %s", clusterName, err)
 	}
 
-	name := api.PackageNamespace
-	key := types.NamespacedName{
-		Name: name,
+	if !tcc.CheckNamespace(ctx, api.PackageNamespace) {
+		err := k8sClient.Create(ctx, &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{Name: api.PackageNamespace},
+		})
+		if err != nil {
+			return fmt.Errorf("create namespace for %q: %s", clusterName, err)
+		}
 	}
-	ns := &corev1.Namespace{}
-	err = k8sClient.Get(ctx, key, ns)
-	// Nil err check here means that the namespace exists thus we can just return with no error
-	if err == nil {
-		return nil
+	return nil
+}
+
+// CheckNamespace tests for the existence of a namespace.
+//
+// It must only be called with an initialized target cluster client.
+func (tcc *targetClusterClient) CheckNamespace(ctx context.Context, namespace string) bool {
+	if tcc.clientConfig == nil {
+		tcc.logger.Error(fmt.Errorf("client is not initialized"), "checking namespace", "namespace", "namespace")
+		return false
 	}
-	if !apierrors.IsNotFound(err) {
-		return fmt.Errorf("get namespace for %s: %s", clusterName, err)
+
+	restConfig, err := tcc.ToRESTConfig()
+	if err != nil {
+		tcc.logger.V(6).Error(err, "creating rest config", "namespace", namespace)
+		return false
 	}
 
-	ns.Name = name
-	err = k8sClient.Create(ctx, ns)
+	k8sClient, err := client.New(restConfig, client.Options{})
```

Comment:
A client is created in the `CreateClusterNamespace`, here and one seems to be part of `tcc` as well.

Reply:
I was minimizing the amount of changes needed to the target cluster client. A follow up to clean that up would be good.